### DISCUSSION TITLE
Checking Balance of a Wallet using wallet_balance_by_index  / wallet_balance_by_address  (Question)

### DISCUSSION THREAD

**Discussion by H-saerdnA** [iotaledger/iota.c#166] [Q&A] [answered]
## Bug description

I have created a wallet with firefly and connected it to the devnet (https://api.lb-0.h.chrysalis-devnet.iota.cafe/). Then i got some Token from online faucet all in all 27 MI.

Then i have created a wallet with iota.c and i used the same Seed passphrase and password like them in the firefly wallet.

Now i checked the balance with iota.c (`wallet_balance_by_index`) and the result was 0MI. I expected the result to be 27MI.

I found out that the function `wallet_balance_by_index` is searching balance from addresses (e.g. atoi1q....5esa5) that are not used in firefly. Ok, then i send to these Adresse (atoi1q....5esa5) 7MI from the Firefly Wallet.

After this i checked the Balance with iota.c (`wallet_balance_by_index)` and i got the Result of 7MI.

And the Balance of the Wallet created with Firefly shows now 20MI. Shouldn't the Firefly Wallet recognize that the address (atoi1q....5esa5) was created from the same seed ?

Then I looked at the two addresses with the balance of 20 MI, used in Firefly ("atoi1q...snwhy" and "atoi1...k05fc").
When I checked the Balance ot these adresses using the functions `address_from_bech32` and `wallet_balance_by_address` from iota.c i got as expected 20MI.

All used Adresses ("atoi1q....5esa5", "atoi1q...snwhy", "atoi1...k05fc") belong to the same Seed but they are created from different wallets (Firefly/IOTA.c).
It seems to be that each wallet searches only the Adresses created by himself.

If I checked the Balance using iota.c (`wallet_balance_by_index`) the result is 7MI.

If I checked the Balance of the Adreses "atoi1q....5esa5", "atoi1q...snwhy" an "atoi1...k05fc" using iota.c (`address_from_bech32` and `wallet_balance_by_address`) the result is a totol of as expected 27MI.

I hope I have described the problem in a reasonably understandable way. 

## Version

Which version of the library are you running?

- Version: 0.2.0

## IOTA network

Which node are you connected to and which IOTA network is it in?

- Node URL: https://api.lb-0.h.chrysalis-devnet.iota.cafe/
- Network: devnet

## Hardware specification

What hardware are you using?

- Operating system: Windows

## Compiler and build tool specification

What compiler are you using? MSVC2019

**Comment by oopsmonk** [comment]
The wallet API is working in progress and not compatible with firefly at the moment. It uses a 32-byte seed to derive addresses, firefly uses a mnemonic sentence and a password, therefore address derivation is different. It will be the same as firefly in next release.

**Comment by H-saerdnA** [comment]
Thank you !

**Comment by H-saerdnA** [comment, marked as answer]
Hello,

I'm testing now the Version 0.3.0. 
Therefore i have created a wallet with iota.c and i use the same Seed passphrase and password like them in my firefly wallet.

In the Dokument "Firefly Wallet Recovery Kit" stands, that the Password is used to generate addresses.

But only if I leave the password free in the function `wallet_create`  I get almost the same results as in Firefly.
```
if ((wallet = wallet_create(D_SEED_PASSPHRASE,  **""**, (int) 0 )) == NULL) {
    return ( FALSE );
  }
```

My question:
is the Password that i used in firefly the same that i have to use in the function `wallet_create` ?


My Test:

1. Create a Wallet with firefly.
2. Got 10 MI from faucet.
3. Create a Wallet with iota.c (same seed  passphrase like in Firefly but with empty pwd)
4. Send 1MI from iota.c-Wallet to an Address created by  firefly.


Here is the address history output of firefly:
```
atoi1qq6kzgyt0v93e6832vr6dqmmpcyyhm2nsd2z0ql6ttgky3p74j3fxz5228h,0
atoi1qz9rf8ly65wvpggd05u8hkhdt6ng2vldt9cpv82vqvhwgyjaadkhz0c2vmx,0
atoi1qzqhlxmxyzhrcllwjc73takygm4ydey60ejykkdvvpqtqrzq4d03ugmhlq0,0
atoi1qppnc8hr4lmprrn3zvplxfc8an8jqpky6rnn9zdyxgwtntsl63hscxk4vwm,1000000
atoi1qrj2x7mzjn4jw6udr5lr76p3jmez78sgps2f25tkwfrh4y8pvy9j70h2pf7,9000000
atoi1qq85waqn2std7j766gd6tjmj93nymk9ls4wwnfmzuw3t9agegqy2czywz8c,0
```


This is the output of the example `wallet_get:_balance` for 10 Adresses:

```
[0] atoi1qrj2x7mzjn4jw6udr5lr76p3jmez78sgps2f25tkwfrh4y8pvy9j70h2pf7: 9000000
[1] atoi1qppnc8hr4lmprrn3zvplxfc8an8jqpky6rnn9zdyxgwtntsl63hscxk4vwm: 1000000
[2] atoi1qzqhlxmxyzhrcllwjc73takygm4ydey60ejykkdvvpqtqrzq4d03ugmhlq0: 0
[3] atoi1qq560dspys2faxy6dmqwwhug5vec8qtgdrjzsjw52ae3upygk8yg7606y7g: 0
[4] atoi1qp8dns6sj2s4620vsjt7xzlklkyqn3swmmudv0ghddka6vqehqv0craammz: 0
[5] atoi1qqcuwme08s7wv2an63ly8qsc4gsk2hgn4xt23rz28kzpw99s4au8uvyxkla: 0
[6] atoi1qzgcdlws2ggey05t9pr0p023egslxy9hwvu5lu77fgn66pummcqa7la5pgh: 0
[7] atoi1qpmmjcy2njhkgfm7l2tskuv670n5e9ajgrw5azlp56utv96kvwa45t3hv7y: 0
[8] atoi1qq7r0x3pcf5txw4qnvt0hzrvwwgwux3635xv2rx9nrw89m03n6u8yeu2uc6: 0
[9] atoi1qp5cjepwhtv558chtfq209qzr8uk3ym2tcpratfalx8x69gfl72aucl8umw: 0
```

The next Address for receiving Token  created by firefly is `atoi1qq6kzgyt0v93e6832vr6dqmmpcyyhm2nsd2z0ql6ttgky3p74j3fxz5228h` but i cannot find this address under which the function `wallet_get_balance` would search.

Can you explain to me how it works or give me a tip where I can read something about it ?

Thanks...

**Reply by oopsmonk** [reply]
> is the Password that i used in firefly the same that i have to use in the function wallet_create ?

No, the Firefly password is used for the Stronghold database. Firefly uses an empty password to derive the mnemonic seed. 
therefore, wallet creation shuld be `wallet_create(eng_mnemonic, "", 0)` where `eng_mnemonic` is an Engligh mnemonic, password is empry `""`, and account index is `0`.

Keep in mind that iota.c is designed for constrained devices, it doesn't provide complicated addresses management like Firefly but provides simple methods for developers to write an address manager for different scenarios.
iota.c and Firefly both follow BIP44 spec, the path is `m/ 44' /4218' /account'/change/address_index`
* account is set via `wallet_create`
* change and address_index are used in `wallet_balance_by_index`

So you need to make sure `account`, `change`, and `address_index` are equal to the Firefly address, please visit [Wallet address management](https://wiki.iota.org/wallet.rs/specification#address)

**Comment by oopsmonk** [comment]
@H-saerdnA dose the description above answer your question?

**Reply by H-saerdnA** [reply]
Yes, thanks for the answer.
The topic is very complex for me and I'm still at the beginning.
But it's very interesting that I can create an address with a high `address_index` in the path and my firefly wallet wouldn't find it. This is how I can hide my Balance :-)
